manpages/imod2vrml2.man
.na
.nh
.TH imod2vrml2 1 2.00 BL3DEMC
.SH NAME
imod2vrml2 \- Make a VRML2.0 model from an IMOD model file.
.SH SYNOPSIS
imod2vrml2  [options]  IMOD_model  WML_file
.SH DESCRIPTION
Convert an IMOD model file to the VRML 2.0 file format used by VRML
viewers and as a common format for interchange of 3D models by
3D animation and CAD programs.  The model will generally 
have similar attributes to those set in 3dmodv before the conversion.
Specifically, objects that are turned off will not
be included in the VRML file.  Scattered points will always display
as shaded spheres ('geometry Sphere' objects), regardless of options
are selected and all meshes should convert across as 'IndexedFaceSet'
objects in the VRML/WML file. If object fill is off points should
be generated using a 'PointSet' object, although these don't
render in all VRML viewers.
.P
For shaded surfaces, the intensity levels may be quite different
in the VRML viewer than in 3dmodv, depending on the settings in the viewer.
If you are generating files for a particular VRML viewer, you can adjust
the Material properties in 3dmodv to get the desired appearance in the viewer.  
The Ambient setting will have no effect, so adjust intensity primarily with
the Diffuse setting, and then with the Specular and Shininess settings.
Transparency may also need to be set higher to achieve the desired level, as
well to minimize artifacts from overlapping transparent
surfaces which become more evident at some orientations.
.SH OPTIONS
.TP 
.B -l
Use low-resolution meshes, if any are available.  The default is to use the
highest resolution meshes contained in the model.  Low resolution meshes can
be produced with the 
.B -l
option to imodmesh(1).
.TP 
.B -a
Output all objects (by default those switched off are omitted).
.TP 
.B -r
Groups all objects together in an 'imod_model' object and rotates this such
that the Z axis become up in Y. This feature is useful when exporting to
animation programs
.TP 
.B -g
Don't assign names to child geometries (by default they get DEF names in
the form 'obj1_mesh' or 'cont2_pt1').
.TP 
.B -n
Output normals (off by default as most 3d programs which import VRML2 can
generate their own normals if missing).
.B -c
Show all contour lines (across all objects) in output VRML file.
.SH AUTHORS
Andrew Noske, based on "imod2vrml" by Jim Kremer and David Mastronarde
.SH SEE ALSO
3dmodv(1), imodmesh(1), imod2vrml(1)
.SH BUGS
Email bug reports to mast@colorado.edu, and CC to andrew<DOT>noske<AT>gmail<DOT>com.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

imodutil/imod2vrml2.c
/*
 *  imod2vrml2 - convert IMOD model to VRML version 2
 *  
 *  Author:     Andrew Noske - modified from James Kremer's imod2vrml
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *  
 *  Info on VRML2: http://www.andrewnoske.com/wiki/index.php?title=VRML
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 *  Log at end
 */

//############################################################

#include <stdio.h>
#include <string.h>
#include <math.h>
#include "imodel.h"

//## GLOBAL VARIABLES AND CONSTANTS:

static int  numObjs     = 0;		// tallies the number of objects turned on
static int  numVertices = 0;		// tallies the number of vertex points added
static int  numFaces    = 0;		// tallies the number of faces added
static int  numSpheres  = 0;		// tallies the number of spheres printed

static int lowRes         = 0;		// if we want to use low res version
static int printAllObjs   = 0;		// print all objects (not just those on)
static int rotateModel    = 0;		// "rotate" (flip Y and Z axis)
static int dontNameGeoms  = 0;		// don't assign names to child geometries
static int printNormals   = 0;		// print normals
static int showContours   = 0;		// show all contour lines

//## FUNCTION DECLARATION:

void imod_to_vrml2(Imod *imod, FILE *fout);

static void printObject(Imod *imod, int ob, FILE *fout);
static void printMaterial(Imod *imod, int ob, int *lastuse, int usefill, FILE *fout);
static void printVrmlSafeUniqueObjectName(Imod *imod, int ob, FILE *fout );
static void printScatContours(Imod *imod, int ob, FILE *fout);
static void printMesh(Imod *imod, int ob, FILE *fout);
static void printMeshPoints(Imesh *mesh, float zscale, FILE *fout);

//############################################################

//------------------------
//-- Usage function - used to write out a small set of instructions in the console
//-- window when the user types "imod2vrml2 -h" or gets an argument wrong.

static void usage(int error)
{
  printf("\nConverts an imod model to the VRML2.0 (Virtual Reality Modeling "
         "Language) file format.\n");
  printf("Usage: imod2vrml2 [options] <imod_model_file.mod> <output_file.wml>\n");
	printf("Options:\n");
  printf("       -l    output low-resolution meshes (if any exist).\n");
	printf("       -a    output all objects (by default those switched off are omitted)\n");
	printf("       -r    groups all objects together in an 'imod_model' object and\n");
	printf("             rotates this such that the Z axis become up in Y\n");
	printf("       -g    don't assign names to child geometries (by default they get \n");
	printf("             DEF names in the form 'obj1_mesh' or 'cont2_pt1').\n");
	printf("       -n    output normals (off by default as most 3d programs which \n");
	printf("             import VRML2 can generate their own normals if missing).\n");
	printf("       -c    show all contour lines.\n");
	
  exit(error);
}


//------------------------
//-- Main function
//-- Reads command line input, opens new file to write out to, prints any
//-- file errors and then calls "imod_to_vrml2" function on the chosen model file.

int main( int argc, char *argv[])
{
  int i;
  FILE *fout;
  Imod *imod;
  char *progname = imodProgName(argv[0]);

  if (argc < 2) {
    imodVersion(progname);
    imodCopyright();
    usage(0);
  }

  for (i = 1; i < argc; i++){
    if (argv[i][0] == '-')
      switch (argv[i][1]){
      case 'l':
        lowRes = 1;
        break;
					
			case 'a':
				printAllObjs = 1;
				break;
			
			case 'r':
				rotateModel = 1;
				break;
			
			case 'g':
				dontNameGeoms = 1;
				break;
				
			case 'n':
				printNormals = 1;
				break;
					
			case 'c':
				showContours = 1;
				break;
				
      default:
        usage(-1);
        break;

      }
    else
      break;
  }

  if (argc - i != 2)
    usage(-1);
	
  imod = imodRead(argv[i]);
  if (!imod){
    fprintf(stderr, "%s: Error reading imod model %s\n", argv[0], 
            argv[i]);
    exit(3);
  }
	
	//## OPEN/CREATE .OBJ FILE FOR WRITING:
	
	fout = fopen( argv[i + 1] , "w");
  if (fout == NULL){
    printf("Couldn't open output file %s.\n", argv[i + 1]);
    exit(10);
  }
	
  imod_to_vrml2(imod, fout);

  fclose(fout);
	
	fprintf(stdout, "Finished writing '%s' \n", argv[i] );
	fprintf(stdout, "  # objects on: %d\n", numObjs     );	
	fprintf(stdout, "  # spheres:    %d\n", numSpheres  );
	fprintf(stdout, "  # vertices:   %d\n", numVertices );
	fprintf(stdout, "  # faces:      %d\n", numFaces    );
  exit(0);
}

//------------------------
//-- Takes an imod model and outputs the meshes into a VRML2 file by printing
//-- then calling "printObject" on each object.
//-- For more information about VRML2 format see:
//-- > http://graphcomp.com/info/specs/sgi/vrml/spec/        (official specs)   OR
//-- > http://www.andrewnoske.com/wiki/index.php?title=VRML  (author's notes)

void imod_to_vrml2(Imod *imod, FILE *fout)
{
	int ob;
  fprintf(fout,"#VRML V2.0 utf8\n");
  fprintf(fout,"#Generated by IMOD\n\n");
	fprintf(fout,"DEF imod_model Transform {\n");
	if(rotateModel)
	  fprintf(fout,"  rotation -1 0 0 1.570796   # rotate by -90 degrees in Z \n");
	fprintf(fout,"  children [\n");
  {
    for(ob = 0; ob < imod->objsize; ob++)
      printObject(imod, ob, fout);
  }
  fprintf(fout,"\n");
	fprintf(fout,"  ]\n");
	fprintf(fout,"}\n");
}

//------------------------
//-- Outputs to file [fout] a unique name for the given object [ob] in the form:
//--   "obj1_Object_Name"
//-- Note that the object number is included (to ensure uniquness) and the user
//-- assigned object name is also printed, but using an underscore (_) to replace
//-- any whitespaces, brackets and several other characters which VRML2.0 format
//-- does not allow in a single name.
//-- These dangerous characters include:    \t \n \\ # c . : () [] {} ' " + & ; |

static void printVrmlSafeUniqueObjectName(Imod *imod, int ob, FILE *fout )
{
	int i;
	Iobj *obj = &imod->obj[ob];
	
	fprintf( fout,"obj%d_", (ob+1) );		// write out object number as "obj1_"
	
	for(i = 0; i < IOBJ_STRSIZE; i++)		// write out object's name as "Object_Name"
	{
		char c = obj->name[i];							// get character
		if( c == 0x00 )
			break;
		else if( c==' ' || c=='\t' || c=='\n' )					// whitespace chars
			fprintf( fout,"_" );
		else if( c=='(' || c=='[' || c=='{' )												// open brackets
			fprintf( fout,"<" );
		else if( c==')' || c==']' || c=='}' )												// close brackets
			fprintf( fout,">" );
		else if( c=='\'' || c=='"' )																// quote symbols
			fprintf( fout,"`" );
		else if( c=='#' || c=='.' || c==',' || c=='\\' || c==':' )	// other
			fprintf( fout,"_" );
		else if( c=='+' || c=='&' || c==';' || c=='|' )			// symbols cinema 4D rejects
			fprintf( fout,"_" );
		else
		  fprintf( fout,"%c", obj->name[i] );
	}
}

//------------------------
//-- Prints out to file the VRML text for that object by first printing
//-- the material by calling "printMaterial" function, then 
//-- calling a seperate function to "printScatContours" or "printMesh"
//-- depending on the way this appeared in code.

static void printObject(Imod *imod, int ob, FILE *fout)
{
  int co;
  Iobj *obj = &imod->obj[ob];
  int lastuse = -1;
	int objScattered = iobjScat(obj->flags) ? 1 : 0;
  int hasSpheres   = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
	if ( !printAllObjs && iobjOff(obj->flags) )		// if object off, and "-a" not entered:
    return;																				// exit early
	
	// See if object has spheres to draw
  for (co = 0; co < obj->contsize && !hasSpheres; co++)
    if (obj->cont[co].sizes)
      hasSpheres = 1;
	
	
	//## PRINT OBJECT'S MATERIAL IN AN EMPTY SHAPE:
	
	printMaterial(imod, ob, &lastuse, obj->flags & IMOD_OBJFLAG_FCOLOR, fout);
	
	
  //## PRINT OBJECT'S DATA:
	
	fprintf(fout,"\n");
	fprintf(fout,"#DATA FOR OBJECT %d:\n", ob+1);
  fprintf(fout,"DEF ");
	printVrmlSafeUniqueObjectName( imod, ob, fout );
	fprintf(fout," Transform {\n");
	fprintf(fout,"  children [\n");
	
	if( !objScattered )
	{
    if (iobjMesh(obj->flags))
      printMesh(imod, ob, fout);
	}
  if (hasSpheres) {
    printScatContours(imod, ob, fout);
  }
	
  fprintf(fout,"  ]\n", ob);
	fprintf(fout,"}\n", ob);
	
  numObjs++;
}


//------------------------
//-- Prints out the VRML2 code representing the material properties of
//-- the specified object ("ob"), matching values as closely as possible.
//-- Things like color and transparency are maintained, but other
//-- values like specular color are not quite perfectly translated.

static void printMaterial(Imod *imod, int ob, int *lastuse, int usefill, FILE *fout)
{
	Iobj *obj = &imod->obj[ob];
  float ambient   = ((int)obj->ambient)   / 255.;
	float diffuse   = ((int)obj->diffuse)   / 255.;
  float specular  = ((int)obj->specular)  / 255.;
  float shininess = ((int)obj->shininess) / 255.;
  float r = obj->red;
  float g = obj->green;
  float b = obj->blue;
  if (*lastuse == usefill)
    return;
  
  if (usefill) {
    r = (float)obj->fillred   / 255.;
    g = (float)obj->fillgreen / 255.;
    b = (float)obj->fillblue  / 255.;
  }
  *lastuse = usefill;
	
  if (shininess < 0.01)
    shininess= 0.01;
	
  //## PRINT EMPTY SHAPE WITH "appearance DEF" TAG SO OTHER SHAPES
	//## CAN REFERENCE THIS TAG:
	
	fprintf(fout,"\n");
	fprintf(fout,"#MATERIAL FOR OBJECT %d:\n", (ob+1) );
  fprintf(fout,"Shape {\n");
  fprintf(fout,"  appearance DEF MAT_");
	printVrmlSafeUniqueObjectName(imod, ob, fout );
	fprintf(fout," Appearance {\n");
	fprintf(fout,"    material Material {\n");
	fprintf(fout,"      ambientIntensity %g \n", ambient );		// "ambientColor" in VRML1
  fprintf(fout,"      diffuseColor %g %g %g\n",  r*diffuse, g*diffuse, b*diffuse);
  fprintf(fout,"      specularColor %g %g %g\n", r*specular, g*specular, b*specular);
  fprintf(fout,"      emissiveColor 0 0 0\n");
  fprintf(fout,"      shininess %g\n", shininess);
  fprintf(fout,"      transparency %g\n", obj->trans/100.0f);
  fprintf(fout,"    }\n");
	fprintf(fout,"  }\n");
	fprintf(fout,"}\n");
}


//------------------------
//-- Prints a scattered point object (in VRML2 format) by printing a sphere for 
//-- every point in the object which has a radius.

static void printScatContours(Imod *imod, int ob, FILE *fout)
{
	Iobj *obj = &imod->obj[ob];
  Icont *cont;
  int hasSpheres = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
  int c, p;
  float zscale = imod->zscale;
  float radius;
  
	fprintf(fout,"    DEF obj%d_SPHERES Group {    # GROUP OF POINTS:\n", ob+1);
	fprintf(fout,"      children [\n\n");	
	
  //## FOR EACH CONTOUR: PRINT EVERY POINT WHICH HAS A SPHERE SIZE:
	
	for(c = 0; c < obj->contsize; c++){
    cont = &(obj->cont[c]);
		
    if (!cont || !cont->psize								// if contour is empty
				|| (!hasSpheres && !cont->sizes))		// or has no sphere sizes:
      continue;																// skip this contour
		
		for (p = 0; p < cont->psize; p++)
		{
      radius = imodPointGetSize(obj, cont, p);
      if (radius > 0.)		// if point radius is non zero: output new sphere
      {
				fprintf(fout,"        DEF obj%d_cont%d_pt%d Transform {\n", ob+1, c+1, p+1 );
				fprintf(fout,"          translation %g %g %g \n",
											      cont->pts[p].x, cont->pts[p].y, (cont->pts[p].z * zscale) );
				if(rotateModel)
					fprintf(fout,"          rotation 1 0 0 1.570796 \n");		// rotate 90 degs
			  fprintf(fout,"          children [ Shape {\n");
				fprintf(fout,"              appearance USE MAT_");
				printVrmlSafeUniqueObjectName(imod, ob, fout );
				fprintf(fout,"\n              ");
				if( dontNameGeoms )
					fprintf(fout,"geometry Sphere { radius %f } } ]\n", radius);
				else
					fprintf(fout,"geometry DEF cont%d_pt%d Sphere { radius %f } } ]\n", 
																									ob+1, c+1, p+1, radius);
				fprintf(fout,"        }\n", radius);
				numSpheres++;
			}
    }
  }
	fprintf(fout,"      ]\n");	
	fprintf(fout,"    }\n");	
}


//------------------------
//-- Prints out the mesh generated over the given object ("ob") in
//-- VRML2 format which involves first outputting the list of all points
//-- using the "printAllPoints" function, then connecting points into triangles.

static void printMesh(Imod *imod, int ob, FILE *fout)
{
  Iobj *obj = &imod->obj[ob];
	int objFill = iobjFill(obj->flags);		// object has been set to show faces
	int objLine = iobjLine(obj->flags);		// object has been set to show lines
  Imesh *mesh;
  int m, c, p, i, j;
  float zscale = imod->zscale;
  int *ilist;
  int lsize;
  int index, ind, poly;
  Ipoint norm;
  int lastuse = -1;
  int resol;
  int listInc, vertBase, normAdd;
	int objClosed, contClosed;
	
	
  imodMeshNearestRes(obj->mesh, obj->meshsize, lowRes, &resol);
  
	
	
	//## FOR EACH MESH (TYPICALLY JUST ONE) PRINT MESH OR POINTS: 
	
  for(m = 0; m < obj->meshsize; m++)
	{
    if (imeshResol(obj->mesh[m].flag) != resol)
      continue;
		
    poly = 0;
    mesh = &obj->mesh[m];
    ilist = (int *)malloc(sizeof(int) * mesh->lsize);
    if (!ilist) {
      printf("Error allocating memory\n");
      exit(1);
    }
		
		
		//## IF OBJECT'S MESH ARE DISPLAYED: GENERATE "IndexFaceSet"
		
		if( objFill )
		{
			//## PRINT START OF NEW SHAPE WITH "IndexFaceSet" GEOMETRY:
			
			fprintf(fout,"    Shape {  #MESH \n");							// start of new Shape
			fprintf(fout,"      appearance USE MAT_" );
			printVrmlSafeUniqueObjectName(imod, ob, fout );
			fprintf(fout,"\n      ");
			if( dontNameGeoms )
				fprintf(fout,"geometry IndexedFaceSet {\n");				// start of new IndexFaceSet
			else
				fprintf(fout,"geometry DEF obj%d_mesh%d IndexedFaceSet {\n", ob+1, 1 );
			fprintf(fout,"        ccw FALSE \n");
			fprintf(fout,"        solid FALSE \n");
			fprintf(fout,"        creaseAngle 1.56207 \n");
			
			//## OUTPUT POINTS:
			printMeshPoints( mesh, zscale, fout );
			
			//## OUTPUT NORMALS (IF SPECIFIED):
			if (printNormals)
			{
				fprintf(fout,"        normal Normal {\n");
				fprintf(fout,"          vector [   # list of all normals for each point\n");
				for(i = 1; i < mesh->vsize; i+=2){
					norm = mesh->vert[i];
					imodPointNormalize(&norm);
					fprintf(fout,"            %.3f %.3f %.3f,\n", norm.x, norm.y, norm.z );
				}
				fprintf(fout,"          ]\n");
				fprintf(fout,"        }\n");
			}
			
			//## OUTPUT ALL MESHES TOGETHER:
			
			fprintf(fout,"        coordIndex [   # connect triangles\n");
			for(i = 0; i < mesh->lsize; i++)
			{
				if (mesh->list[i] == IMOD_MESH_END)
					break;
				if ( !imodMeshPolyNormFactors(mesh->list[i], &listInc, &vertBase, &normAdd) )
					continue;
				
				lsize = 0;		// copy the indices to the new index array numbered from 0:
				i++;
				while (mesh->list[i] != IMOD_MESH_ENDPOLY)
				{
					ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
					ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
					ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
				}
				
				if (!lsize)
					continue;
				
				//## OUTPUT FACESET:   (if showing surface, put out face set)
				
				for(index = 0; index < lsize / 3; index++)
				{
					ind = 3 * index;
					fprintf(fout,"          %d,%d,%d,-1,\n", 
									ilist[ind+2], ilist[ind+1], ilist[ind]);
					numFaces++;
				}
			}
			fprintf(fout,"        ]\n");
			
			//## PRINT END OF SHAPE:
			
			fprintf(fout,"      }\n");		// end of IndexFaceSet
			fprintf(fout,"    }\n");			// end of Shape
		}
		
		
		//## IF OBJECT RENDERED AS POINTS: GENERATE "PointSet"
		
		if( !objFill )
		{
			//## PRINT START OF NEW SHAPE WITH "IndexFaceSet" GEOMETRY:
			
			fprintf(fout,"    Shape {\n");							// start of new Shape
			fprintf(fout,"      appearance USE MAT_" );
			printVrmlSafeUniqueObjectName(imod, ob, fout );
			fprintf(fout,"\n      ");
			if( dontNameGeoms )
				fprintf(fout,"geometry PointSet {\n");				// start of new PointSet
			else
				fprintf(fout,"geometry DEF obj%d_all_points PointSet {\n", ob+1, 1 );
			
			//## OUTPUT POINTS:
			printMeshPoints( mesh, zscale, fout );
			
			fprintf(fout,"      }\n");		// end of PointSet
			fprintf(fout,"    }\n");			// end of Shape
			
    }
		
		
		//## FREE ANY MEMORY USED:
    if (ilist) 
      free(ilist);
  }
	
	
	//## IF OBJECT'S CONTOURS ARE DISPLAYED: GENERATE "IndexLineSet"
	
	if (showContours)
	{
		objClosed = (imodObjectGetValue(obj,IobjFlagClosed)==1) ? 1 : 0;
		for(c = 0; c < obj->contsize; c++)
		{
			Icont *cont = &obj->cont[c];
			if(cont->psize == 0)
				continue;
			
			//## PRINT START OF NEW SHAPE WITH "IndexFaceSet" GEOMETRY:
			
			fprintf(fout,"    Shape {  #CONTOUR \n");							// start of new Shape
			fprintf(fout,"      appearance USE MAT_" );
			printVrmlSafeUniqueObjectName(imod, ob, fout );
			fprintf(fout,"\n      ");
			if( dontNameGeoms )
				fprintf(fout,"geometry IndexedLineSet {\n");				// start of new IndexFaceSet
			else
				fprintf(fout,"geometry DEF obj%d_cont%d IndexedLineSet {\n", ob+1, c+1 );
			
			//## PRINT POINTS:
			
			contClosed = (objClosed || imodContourGetFlag(cont,ICONT_OPEN) != 0) ?1:0;
			
			fprintf(fout,"        coord Coordinate {\n");
			fprintf(fout,"          point [   # contour points\n");
			for(p = 0; p < cont->psize; p++)
			{
				fprintf(fout,"            %.5g %.5g %.5g,\n",
								cont->pts[p].x, cont->pts[p].y, (cont->pts[p].z * zscale) );
				
				numVertices++;
			}
			fprintf(fout,"          ]\n");
			fprintf(fout,"        }\n");
			fprintf(fout,"        coordIndex [     # connect points\n");
			fprintf(fout,"          \n");
			for(p = 0; p < cont->psize; p++)
				fprintf(fout,"%d,", p);
			if(contClosed)
				fprintf(fout,"0");
			fprintf(fout,"        ]\n");
			fprintf(fout,"      }\n");		// end of IndexFaceSet
			fprintf(fout,"    }\n");			// end of Shape
		}
		
	}
	
	
}

//------------------------
//-- Prints a list of all points in the specified mesh in VRML2 format
//-- as an point array in the form:
//-- "coord Coordinate { point [ 0 0 0, 0 0 5, ... ] )".

static void printMeshPoints(Imesh *mesh, float zscale, FILE *fout)
{
	int i;
	fprintf(fout,"        coord Coordinate {\n");
	fprintf(fout,"          point [   # list of all points in mesh\n");
	for(i = 0; i < mesh->vsize; i+=2)
	{
		fprintf(fout,"            %.5g %.5g %.5g,\n",
						mesh->vert[i].x, mesh->vert[i].y, (mesh->vert[i].z * zscale) );
		
		numVertices++;
	}
	fprintf(fout,"          ]\n");
	fprintf(fout,"        }\n");
}


/*
$Log: imod2vrml2.c,v $
Revision 1.0  2011/05/29 21:37:14  noske
Modified file from "imod2vrml.c" to output VRML2.0 format
*/
